python/ambassador/ir/irutils.py

```diff
@@ -17,32 +17,48 @@
 import logging
 
 ################
-## hostglob_matches is a utility for host globbing.
-
-def hostglob_matches(glob: str, value: str) -> bool:
-    """
-    Does a host glob match a given value?
-    """
-
-    rc = False
-
-    if ('*' in value) and not ('*' in glob):
-        # Swap.
-        tmp = value
-        value = glob
-        glob = tmp
-
-    if glob == "*": # special wildcard
-        rc=True
-    elif glob.endswith("*"): # prefix match
-        rc=value.startswith(glob[:-1])
-    elif glob.startswith("*"): # suffix match
-        rc=value.endswith(glob[1:])
-    else: # exact match
-        rc=value == glob
-
-    # sys.stderr.write(f"hostglob_matches: {value} gl~ {glob} == {rc}\n")
-    return rc
+## hostglob_matches determines two given DNS globs fit each other,
+## in either direction.
+
+def hostglob_matches(g1: str, g2: str) -> bool:
+    # Special case: if either glob is "*", then it matches everything.
+    if (g1 == "*") or (g2 == "*"):
+        return True
+
+    # OK, we don't have the simple-"*" case, so any wildcards must be at
+    # the start or end, and they must be a component alone.
+    if not (g1.startswith("*.") or g1.endswith("*") or
+            g2.startswith("*.") or g2.endswith("*")):
+        # OK, no valid wildcards, do a simple equality check.
+        return g1 == g2
```

Comment:
This doesn't allow partial-label prefix wildcards like `*-bar.foo.com`, which Envoy allows.

See:
- https://www.envoyproxy.io/docs/envoy/v1.17.3/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-virtualhost-domains
- https://github.com/datawire/envoy/blob/ambassador-1.14.1/api/envoy/config/route/v3/route_components.proto#L65-L86
- https://github.com/datawire/envoy/blob/ambassador-1.14.1/source/common/router/config_impl.cc#L1280-L1305